include/ccf/indexing/strategies/seqnos_by_key_bucketed.h

```diff
@@ -0,0 +1,373 @@
+        }
+      }
+
+      SeqNoCollection result;
+
+      auto append_bucket_result = [&](const SeqNoCollection& seqnos) {
+        for (auto n : seqnos)
+        {
+          if (n >= from && n <= to)
+          {
+            result.insert(n);
+          }
+        }
+      };
+
+      bool complete = true;
+
+      while (true)
+      {
+        const auto bucket_key = std::make_pair(serialised_key, from_range);
+        const auto old_it = old_results.find(bucket_key);
+        if (old_it != old_results.end())
+        {
+          auto& bucket_value = old_it->second;
+
+          // We were already trying to fetch this. If it's finished fetching,
+          // parse and store the result
+          if (bucket_value.first != nullptr)
+          {
+            const auto fetch_result = bucket_value.first->fetch_result;
+            switch (fetch_result)
+            {
+              case (FetchResult::Fetching):
+              {
+                complete = false;
+                break;
+              }
+              case (FetchResult::Loaded):
+              {
+                bool corrupt = false;
+                bucket_value.second =
+                  deserialise(bucket_value.first->contents, corrupt);
+                if (!corrupt)
+                {
+                  bucket_value.first = nullptr;
+                  break;
+                }
+                else
+                {
+                  // Deliberately fall through to the case below. If this can't
+                  // deserialise the value, consider the file corrupted
+                  LOG_FAIL_FMT("Deserialisation failed");
+                }
+              }
+              case (FetchResult::NotFound):
+              case (FetchResult::Corrupt):
+              {
+                // This class previously wrote a bucket to disk which is no
+                // longer available or corrupted. Reset the watermark of what
+                // has been indexed, to re-index and rewrite those files.
+                complete = false;
+                const auto problem =
+                  fetch_result == FetchResult::NotFound ? "missing" : "corrupt";
+                LOG_FAIL_FMT(
+                  "A file that {} requires is {}. Re-indexing.",
+                  get_name(),
+                  problem);
+                LOG_DEBUG_FMT(
+                  "The {} file is {}", problem, bucket_value.first->key);
+
+                // NB: This could probably be more precise about what is
+                // re-indexed. Technically only need to build an index to build
+                // the current query, and that applies to re-indexing too. But
+                // for safety, and consistency with the simple indexing
+                // strategies currently used, this re-indexes everything from
+                // the start of time.
+                current_txid = {};
+                old_results.clear();
+                current_results.clear();
```

Comment:
Related to the comment on `strategy.h`, this is where we begin to re-index if we were given a corrupt file. Note that as the comment says, we're pessimistic and eager here; we re-index everything from 0, where we could technically (probably) just re-index to answer the current query (lazy re-indexing). But avoiding such complications for now, and accepting resulting inefficiency.